### PR TITLE
Add grammar documentation to all relevant keys

### PR DESCRIPTION
**Pull request by kyrofa**
This currently includes `stage-packages`, `build-packages`, and `source`.

### CODE REVIEW

File: build-snaps/syntax.md [hunk 1]
```diff
@@ -449,6 +449,23 @@ directory tree or a tarball or a revision control repository
           plugin: dump
           source: bar-0.1.tar
 
+As in the example, one may simply specify a string, in which case that source
+will be used regardless of the build environment.
+
+If the source instead needs to be dependent on the build environment, this
```
Comment (caldav) [inline build-snaps/syntax.md]: "if the sources needs to be dependent"

File: build-snaps/syntax.md [hunk 2]
```diff
@@ -449,6 +449,23 @@ directory tree or a tarball or a revision control repository
...
+keywords supports Snapcraft's [advanced grammar](#advanced-grammar). For
```
Comment (caldav) [inline build-snaps/syntax.md]: keyword

File: build-snaps/syntax.md [hunk 3]
```diff
@@ -449,6 +449,23 @@ directory tree or a tarball or a revision control repository
...
+example, if you require a specific source to be used on amd64, and another for
```
Comment (caldav) [inline build-snaps/syntax.md]: "for example, you can require a specific source"

File: build-snaps/syntax.md [hunk 4]
```diff
@@ -870,3 +901,52 @@ A list of special attributes that affect the build of this specific part:
         foo:
           plugin: kbuild
           build-attributes: [debug, no-system-libraries]
+
+## Advanced grammar
+
+Several fields in the YAML support being architecture- or
```
[the remark this answers is not in the document]
Comment (kyrofa) [reply build-snaps/syntax.md]: That's a little forward-looking, as the grammar only supports architectures right now. Think it's still worth adding?

[the remark this answers is not in the document]
Comment (kyrofa) [reply build-snaps/syntax.md]: Fixed.

**Pull request merged**